### PR TITLE
Redirect function for HTTP Server Connector

### PR DESCRIPTION
**Pull request by irunika**
## Purpose
ATM users can manually define redirect responses in any HTTP server connector code but it is better is we can provide more systematic way to do this.
This PR resolves https://github.com/ballerina-lang/ballerina/issues/4246

## Goals
To ease the way of responding with redirection details.

## Approach
Add method to http: Connection struct.
```ballerina
public function <Connection conn> redirect(OutResponse response, RedirectCode code, string[] locations) (HttpConnectorError)
```
## Automation tests
 - Unit tests 
   >Yes
 - Integration tests
   > No
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
```ballerina
import ballerina.net.http;

@http:configuration {
    basePath:"/hello"
}
service<http> helloServer {
    @http:resourceConfig {
        path:"/redirect",
        methods:["GET"]
    }
    resource redirect(http:Connection conn, http:InRequest req) {
        http:OutResponse res = {};
        _ = conn.redirect(res, http:RedirectCode.MOVED_PERMANENTLY_301, ["location1"]);
    }
}
```

### CODE REVIEW

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 1]
```diff
@@ -1,5 +1,17 @@
...
+const string HEADER_KEY_LOCATION = "Location";
+
+public enum redirectCode {
+    MULTIPLE_CHOICES_300,
+    MOVED_PERMANENTLY_301,
+    FOUND_302,
+    SEE_OTHER_303,
+    NOT_MODIFIED_304,
+    USE_PROXY_305,
+    TEMPORARY_REDIRECT_307
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Need to document these enum fields.

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 2]
```diff
@@ -9,3 +21,37 @@ public function <Connection conn> respondContinue () (HttpConnectorError) {
     HttpConnectorError err = conn.respond(res);
     return err;
 }
+
+@Description { value:"Sends a redirect response code to the user with given redirection status code." }
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: "code" in "redirect response code" sounds outplace here.. Shall we just say redirect response?

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 3]
```diff
@@ -9,3 +21,37 @@ public function <Connection conn> respondContinue () (HttpConnectorError) {
...
+@Param { value:"conn: The server connector connection" }
+@Param { value:"response: Response which user wants to send." }
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Shall we rephrase this to something like "Response to be sent to the client"?

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal (hunk 1 above)
Comment (pubudu91) [reply stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Also, name of the enum should be `RedirectCode`

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 4]
```diff
@@ -9,3 +21,37 @@ public function <Connection conn> respondContinue () (HttpConnectorError) {
...
+        response.statusCode = 302;
+    } else if (code == redirectCode.SEE_OTHER_303) {
+        response.statusCode = 303;
+    } else if (code == redirectCode.NOT_MODIFIED_304) {
+        response.statusCode = 304;
+    } else if (code == redirectCode.USE_PROXY_305) {
+        response.statusCode = 305;
+    } else if (code == redirectCode.TEMPORARY_REDIRECT_307) {
+        response.statusCode = 307;
+    }
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Considering the lengthy `if-else` clauses required, does the use of enums for the status code seem justified? Wouldn't just integer constants be better suited? Then we can just get rid of the conditional checks.

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 5]
```diff
@@ -9,3 +21,37 @@ public function <Connection conn> respondContinue () (HttpConnectorError) {
...
     return err;
 }
+
+@Description { value:"Sends a redirect response code to the user with given redirection status code." }
+@Param { value:"conn: The server connector connection" }
+@Param { value:"response: Response which user wants to send." }
+@Param { value:"redirectCode: Status code of the specific redirect." }
+@Param { value:"locations: Array of locations where the redirection can happen." }
+@Return { value:"Returns an HttpConnectorError if there was any issue in sending the response." }
+public function <Connection conn> redirect(OutResponse response, redirectCode code, string[] locations) (HttpConnectorError) {
```
Comment (chamil321) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Shall we add a space after "redirect" ?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/nativeimpl/connection/ConnectionNativeFunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import org.ballerinalang.launcher.util.CompileResult;
+import org.ballerinalang.net.http.HttpConstants;
+import org.ballerinalang.test.services.testutils.HTTPTestRequest;
+import org.ballerinalang.test.services.testutils.MessageUtils;
+import org.ballerinalang.test.services.testutils.Services;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import org.wso2.transport.http.netty.message.HTTPCarbonMessage;
+
```
Comment (chamil321) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/services/nativeimpl/connection/ConnectionNativeFunctionTest.java]: Shall we add the class description?

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 7]
```diff
@@ -9,3 +21,37 @@ public function <Connection conn> respondContinue () (HttpConnectorError) {
...
+    string locationsStr = "";
+    foreach location in locations {
+        locationsStr = locationsStr + location + ",";
+    }
+    locationsStr = locationsStr.subString(0, (lengthof locationsStr) - 1);
+
+    response.setHeader(HEADER_KEY_LOCATION, locationsStr);
+    var e = conn.respond(response);
+    return e;
+}
```
Comment (chamil321) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Seems this file is not ended with a new line?

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal (hunk 4 above)
Comment (irunika) [reply stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Yes. That is true. But my point is that there are only those given cases when redirecting. So giving an opportunity for the user to set any code is not appropriate. IMO either we use this way or give integer constant, we need to check whether the correct status code is set for the redirect.

Comment (pubudu91) [reply stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Well, you could have one conditional check to ensure the provided status code falls within the permissible values for redirect status codes.

Alternatively, if we had a way to add fields inside an enum (like in Java), this could be solved neatly using enums.

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/nativeimpl/connection/ConnectionNativeFunctionTest.java (hunk 6 above)
Comment (irunika) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/services/nativeimpl/connection/ConnectionNativeFunctionTest.java]: Done

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal (hunk 7 above)
Comment (irunika) [reply stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Done

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 8]
```diff
@@ -1,5 +1,25 @@
...
+const string HEADER_KEY_LOCATION = "Location";
+
+@Description { value:"Status codes for HTTP redirect"}
+@Field { value:"MULTIPLE_CHOICES_300: Represent status code 300 - Mutlipe choises."}
+@Field { value:"MOVED_PERMANENTLY_301: Represent status code 301 - Moved Permanantly."}
+@Field { value:"FOUND_302: Represent status code 302 - Found."}
+@Field { value:"SEE_OTHER_303: Represent status code 303 - See Other."}
+@Field { value:"NOT_MODIFIED_304: Represent status code 304 - Not Modified."}
+@Field { value:"USE_PROXY_305: Represent status code 305 - Use Proxy."}
+@Field { value:"TEMPORARY_REDIRECT_307: Represent status code 307 - Temporary Redirect."}
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Please double check these doc comments.. There are typos

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 9]
```diff
@@ -9,3 +29,37 @@ public function <Connection conn> respondContinue () (HttpConnectorError) {
...
+    }
+
+    string locationsStr = "";
+    foreach location in locations {
+        locationsStr = locationsStr + location + ",";
+    }
+    locationsStr = locationsStr.subString(0, (lengthof locationsStr) - 1);
+
+    response.setHeader(HEADER_KEY_LOCATION, locationsStr);
+    var e = conn.respond(response);
```
Comment (anupama-pathirage) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: var e is redundant here and can we directly return ?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/nativeimpl/connection/ConnectionNativeFunctionTest.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Test cases for ballerina.net.http.response native functions.
+ */
+public class ConnectionNativeFunctionTest {
+
+    private CompileResult serviceResult;
+    private String filePath =
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/services/nativeimpl/connection/ConnectionNativeFunctionTest.java]: Shall we make this a local variable?

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal (hunk 9 above)
Comment (shafreenAnfar) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Shall we not use single letter variables please..

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal (hunk 1 above)
Comment (irunika) [reply stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Done
